av.h

```diff
@@ -112,64 +112,96 @@ If all you need is to look up an array element, then prefer C<av_fetch>.
 =for apidoc_item newAV_alloc_x
 =for apidoc_item newAV_alloc_xz
 
-These all create a new AV, setting the reference count to 1. They differ
-in the allocation and population of the array of SV*s that always
-accompanies a non-empty AV.
+These all create a new AV, setting the reference count to 1.
 
-The Perl equivalent is approximately C<my @array;>.
+As background, an array consists of three things:
 
-newAV does not allocate a SV* array.
-    AV *av = newAV();
+=over
+
+=item 1.
+
+A data structure containing information about the array as a whole, such as its
+size and reference count.
+
+=item 2.
+
+A C language array of pointers to the individual elements.  These are treated
+as pointers to SVs, so all must be castable to SV*.
+
+=item 3.
+
+The individual elements themselves.  These could be, for instance, SVs and/or
+AVs and/or HVs, etc.
+
+=back
+
+An empty array need only have the first data structure, and all these functions
+create that.  They differ in what else they do, as follows:
+
+=over
+
+=item C<newAV> form
+
+=for comment
+'form' above and below is because otherwise have two =items with the same name,
+can't link to them.
+
+This does nothing beyond creating the whole-array data structure.
+The Perl equivalent is approximately S<C<my @array;>>
 
-This is very useful when an AV is required, but populating it may be
-deferred, or even never actually take place. (Memory is not allocated
-unnecessarily.)
+This is useful when the minimum size of the array could be zero (perhaps there
+are likely code paths that will entirely skip using it).
 
-Subsequent SV* array allocation would be performed via C<L</av_extend>>.
-This might be called directly:
-    av_extend(av, key);
+If the array does get used, the pointers data structure will need to be
+allocated at that time.  This will end up being done by L</av_extend>>,
+either explicitly:
+
+    av_extend(av, len);
+
+or implicitly when the first element is stored:
 
-Or it might be called implicitly when the first element is stored:
     (void)av_store(av, 0, sv);
 
 Unused array elements are typically initialized by C<av_extend>. (Only
 core maintainers should have need to concern themseleves with when that
-is not the case. Refer to F<av.h> and F<av.c> for the differences between
+is not the case.
+
+Refer to F<av.h> and F<av.c> for the differences between
```

Comment:
Ok, I removed the spacing.   I didn't understand this point.  Do we need to even have mention in the API about things that don't apply to most people?  Could this be in a regular comment somewhere, like after the =cut.  I don't know, and am deferring to you

Reply:
Yes, I think the "Refer to" sentence could be a regular comment. It doesn't seem necessary for it to be in the apidoc, as it could unnecessarily confuse, but I like having it somewhere for newbie porters (like me) to follow if they want to learn the subtleties of AVs in core.

Reply:
done